### PR TITLE
Add X3d models

### PR DESCRIPTION
**Pull request by kennymckormick**
In this PR, code of X3d and the models ported from SlowFast are added.

### CODE REVIEW

File: configs/recognition/x3d/README.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+[1] The models are ported from the repo [SlowFast](https://github.com/facebookresearch/SlowFast/) and tested on our data. Currently, we only support the testing of X3d models, training will be available soon.
+
+Notes:
+
+1. The **gpus** indicates the number of gpu we used to get the checkpoint. It is noteworthy that the configs we provide are used for 8 gpus as default.
+According to the [Linear Scaling Rule](https://arxiv.org/abs/1706.02677), you may set the learning rate proportional to the batch size if you use different GPUs or videos per GPU,
+e.g., lr=0.01 for 4 GPUs * 2 video/gpu and lr=0.08 for 16 GPUs * 4 video/gpu.
+2. The **inference_time** is got by this [benchmark script](/tools/analysis/benchmark.py), where we use the sampling frames strategy of the test setting and only care about the model inference time,
+not including the IO time and pre-processing time. For each setting, we use 1 gpu and set batch size (videos per gpu) to 1 to calculate the inference time.
```
Comment (dreamerlin) [inline configs/recognition/x3d/README.md:32]: Notes 1 and 2 can be removed since they are not shown in the table.

File: configs/recognition/x3d/README.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+```
+
+## Model Zoo
+
+### Kinetics-400
+
+|config | resolution | backbone | top1 10-view | top1 30-view | reference top1 10-view | reference top1 30-view | ckpt |
+|:--|:--:|:--:|:--:|:--:|:--:|:--:|:--:|
+|[x3d_s_13x6x1_facebook_kinetics400_rgb](/configs/recognition/x3d/x3d_s_13x6x1_facebook_kinetics400_rgb.py)|short-side 320| X3D_S | 72.7 | 73.2 | [73.1](https://github.com/facebookresearch/SlowFast/blob/master/MODEL_ZOO.md) | [73.5](https://github.com/facebookresearch/SlowFast/blob/master/MODEL_ZOO.md) | [ckpt]()[1] |
+|[x3d_m_16x5x1_facebook_kinetics400_rgb](/configs/recognition/x3d/x3d_m_16x5x1_facebook_kinetics400_rgb)|short-side 320| X3D_M | 75.0 | 75.6 | [75.1](https://github.com/facebookresearch/SlowFast/blob/master/MODEL_ZOO.md) | [76.2](https://github.com/facebookresearch/SlowFast/blob/master/MODEL_ZOO.md) | [ckpt]()[1] |
```
Comment (dreamerlin) [inline configs/recognition/x3d/README.md:22]: Use the url with the hash tag for reference repo

File: configs/recognition/x3d/x3d_s_13x6x1_facebook_kinetics400_rgb.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    dict(
+        type='SampleFrames',
+        clip_len=13,
+        frame_interval=6,
+        num_clips=10,
+        test_mode=True),
+    dict(type='RawFrameDecode'),
+    dict(type='Resize', scale=(-1, 192)),
+    dict(type='CenterCrop', crop_size=192),
+    dict(type='Flip', flip_ratio=0),
```
Comment (dreamerlin) [inline configs/recognition/x3d/x3d_s_13x6x1_facebook_kinetics400_rgb.py:31]: This can be removed to save time.

File: configs/recognition/x3d/x3d_m_16x5x1_facebook_kinetics400_rgb.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    dict(
+        type='SampleFrames',
+        clip_len=16,
+        frame_interval=5,
+        num_clips=10,
+        test_mode=True),
+    dict(type='RawFrameDecode'),
+    dict(type='Resize', scale=(-1, 256)),
+    dict(type='CenterCrop', crop_size=256),
+    dict(type='Flip', flip_ratio=0),
```
Comment (dreamerlin) [inline configs/recognition/x3d/x3d_m_16x5x1_facebook_kinetics400_rgb.py:31]: Remove this step

File: mmaction/datasets/pipelines/augmentations.py [hunk 5]
```diff
@@ -533,6 +533,85 @@ def __repr__(self):
         return repr_str
 
 
+@PIPELINES.register_module()
+class RandomRescale(object):
```
Comment (dreamerlin) [inline mmaction/datasets/pipelines/augmentations.py:537]: ~(object)~

File: mmaction/models/backbones/x3d.py [hunk 6]
```diff
@@ -0,0 +1,521 @@
...
+            planes (int): Number of channels for the output feature
+                in each block. Base_channel * gamma_w * gamma_b.
+            blocks (int): Number of residual blocks.
+            spatial_stride (int): Spatial strides in
+                residual and conv layers. Default: 1.
+            se_style (str): The style of inserting SE modules into BlockX3d,
+                'half' denotes insert into half of the blocks, while 'all'
+                denotes insert into all blocks. Default: 'half'.
+            se_ratio (float): The reduction ratio of squeeze and excitation
+                unit. `se_ratio == None` indicates not using SE unit.
```
[the remark this answers is not in the document]
Comment (dreamerlin) [reply mmaction/models/backbones/x3d.py:366]: It is a gramma different from markdown, I'm sorry about forgetting the name : )

File: mmaction/models/heads/x3d_head.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+    Args:
+        num_classes (int): Number of classes to be classified.
+        in_channels (int): Number of channels in input feature.
+        loss_cls (dict): Config for building loss.
+            Default: dict(type='CrossEntropyLoss')
+        spatial_type (str): Pooling type in spatial dimension. Default: 'avg'.
+        dropout_ratio (float): Probability of dropout layer. Default: 0.5.
+        init_std (float): Std value for Initiation. Default: 0.01.
+    """
```
Comment (dreamerlin) [inline mmaction/models/heads/x3d_head.py:20]: Docstring here is outdated.

File: mmaction/datasets/pipelines/augmentations.py [hunk 8]
```diff
@@ -533,6 +533,85 @@ def __repr__(self):
...
+        assert len(scale_range) == 2
+        assert type(scale_range[0]) is int
+        assert type(scale_range[1]) is int
+        assert scale_range[0] < scale_range[1]
+
+        self.keep_ratio = True
+        self.interpolation = interpolation
+        self.lazy = lazy
+
+    def __call__(self, results):
```
[the remark this answers is not in the document]
Comment (kennymckormick) [reply mmaction/datasets/pipelines/augmentations.py:566]: I'm not sure. If constructing a new `Resize` instance is not time-consuming, it will be OK.

File: mmaction/models/backbones/x3d.py [hunk 9]
```diff
@@ -0,0 +1,521 @@
...
+# isort:end_maintain_block
+
+
+class SEModule(nn.Module):
+
+    def __init__(self, channels, reduction):
+        super(SEModule, self).__init__()
+        self.avg_pool = nn.AdaptiveAvgPool3d(1)
+        self.bottleneck = self._round_width(channels, reduction)
+        self.fc1 = nn.Conv3d(
```
[the remark this answers is not in the document]
Comment (kennymckormick) [reply mmaction/models/backbones/x3d.py:25]: I think calling it `fc` is better: 1. The layer is actually an `fc` layer, the feature map it operates on is of the size 1x1x1. Using `conv3d` to implement it can save 2 additional reshape. 2. In both SE-ResNet and the SlowFast repo, people call it `fc`.

File: mmaction/models/backbones/x3d.py [hunk 10]
```diff
@@ -0,0 +1,521 @@
...
+                 **kwargs):
+        super().__init__()
+        self.gamma_w = gamma_w
+        self.gamma_b = gamma_b
+        self.gamma_d = gamma_d
+
+        self.pretrained = pretrained
+        self.in_channels = in_channels
+        # Hard coded, can be changed by gamma_w
+        self.base_channels = 24
```
[the remark this answers is not in the document]
Comment (kennymckormick) [reply mmaction/models/backbones/x3d.py:240]: It would be redundant. Actually, the `base_channels` of the X3D model equals to self.base_channel * self.gamma_w

File: mmaction/models/backbones/x3d.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kennymckormick) [reply mmaction/models/backbones/x3d.py:368]: The docstring is fixed. BTW, what's the meaning of the second comment?

File: mmaction/models/backbones/x3d.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dreamerlin) [reply mmaction/models/backbones/x3d.py:25]: Argee

File: mmaction/datasets/pipelines/augmentations.py [hunk 11]
```diff
@@ -533,6 +533,85 @@ def __repr__(self):
         return repr_str
 
 
+@PIPELINES.register_module()
+class RandomRescale:
```
Comment (dreamerlin) [inline mmaction/datasets/pipelines/augmentations.py:537]: This class seems not unittested

File: mmaction/datasets/pipelines/augmentations.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (innerlee) [reply mmaction/datasets/pipelines/augmentations.py:566]: yeah there are some overhead, can this object be moved to init?

[the remark this answers is not in the document]
Comment (kennymckormick) [reply mmaction/datasets/pipelines/augmentations.py:566]: No, since the target size is dynamically determined.

[the remark this answers is not in the document]
Comment (dreamerlin) [reply mmaction/datasets/pipelines/augmentations.py:563]: How about initializing the `Resize` in `def __init__`, but changing the `resize.scale` in `__call__`.

[the remark this answers is not in the document]
Comment (kennymckormick) [reply mmaction/datasets/pipelines/augmentations.py:563]: I think ur first version would be nice. Although it may introduce additional overhead, it simplifies the code and doesn't require explicitly setting attributes of an existing object. BTW, the overhead is very little comparing to the data loading pipeline.

[the remark this answers is not in the document]
Comment (innerlee) [reply mmaction/datasets/pipelines/augmentations.py:563]: okay makes sense

File: mmaction/models/backbones/x3d.py [hunk 12]
```diff
@@ -0,0 +1,526 @@
...
+from ..registry import BACKBONES
+
+# pylint: enable=no-name-in-module
+# isort:end_maintain_block
+
+
+class SEModule(nn.Module):
+
+    def __init__(self, channels, reduction):
+        super(SEModule, self).__init__()
```
Comment (innerlee) [inline mmaction/models/backbones/x3d.py:22]: `super().__init__()`

File: mmaction/models/backbones/x3d.py [hunk 13]
```diff
@@ -0,0 +1,526 @@
+# pylint: disable=no-name-in-module
+# isort:maintain_block
```
Comment (innerlee) [inline mmaction/models/backbones/x3d.py:2]: what was wrong with isort?

**Review by innerlee — APPROVED** [review]

File: mmaction/models/backbones/x3d.py (hunk 13 above)
Comment (kennymckormick) [reply mmaction/models/backbones/x3d.py]: Now it's OK to remove these lines.

**Comment by innerlee** [comment]
conflicts
